server/src/test/java/org/elasticsearch/cluster/routing/allocation/RetryFailedAllocationTests.java

```diff
@@ -0,0 +1,101 @@
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.ShardRoutingState;
+import org.elasticsearch.cluster.routing.allocation.command.AllocateReplicaAllocationCommand;
+import org.elasticsearch.cluster.routing.allocation.command.AllocationCommands;
+import org.elasticsearch.cluster.routing.allocation.decider.MaxRetryAllocationDecider;
+import org.elasticsearch.common.settings.Settings;
+
+import java.util.Collections;
+import java.util.List;
+
+public class RetryFailedAllocationTests extends ESAllocationTestCase {
+
+    private MockAllocationService strategy;
+    private ClusterState clusterState;
+    private final String INDEX_NAME = "index";
+
+    @Override
+    public void setUp() throws Exception {
+        super.setUp();
+        MetaData metaData = MetaData.builder().put(IndexMetaData.builder(INDEX_NAME)
+            .settings(settings(Version.CURRENT)).numberOfShards(1).numberOfReplicas(1)).build();
+        RoutingTable routingTable = RoutingTable.builder().addAsNew(metaData.index(INDEX_NAME)).build();
+        clusterState = ClusterState.builder(ClusterName.DEFAULT).metaData(metaData).routingTable(routingTable)
+            .nodes(DiscoveryNodes.builder().add(newNode("node1")).add(newNode("node2"))).build();
+        strategy = createAllocationService(Settings.EMPTY);
+    }
+
+    private ShardRouting getPrimary() {
+        for (ShardRouting shard : clusterState.getRoutingTable().allShards()) {
+            if (shard.getIndexName().equals(INDEX_NAME) && shard.primary()) {
+                return shard;
+            }
+        }
+        throw new IllegalArgumentException("No primary found for index: " + INDEX_NAME);
+    }
+
+    private ShardRouting getReplica() {
+        for (ShardRouting shard : clusterState.getRoutingTable().allShards()) {
+            if (shard.getIndexName().equals(INDEX_NAME) && !shard.primary()) {
+                return shard;
+            }
+        }
+        throw new IllegalArgumentException("No replica found for index: " + INDEX_NAME);
+    }
+
+    public void testRetryFailedResetForAllocationCommands() {
+        final int retries = MaxRetryAllocationDecider.SETTING_ALLOCATION_MAX_RETRY.get(Settings.EMPTY);
+        clusterState = strategy.reroute(clusterState, "initial allocation");
+        clusterState = strategy.applyStartedShards(clusterState, Collections.singletonList(getPrimary()));
+
+        // Exhaust all replica allocation attempts with shard failures
+        for (int i = 0; i < retries; i++) {
+            List<FailedShard> failedShards = Collections.singletonList(
+                new FailedShard(getReplica(), "failing-shard::attempt-" + i,
+                    new UnsupportedOperationException(), randomBoolean()));
+            clusterState = strategy.applyFailedShards(clusterState, failedShards);
+            clusterState = strategy.reroute(clusterState, "allocation retry attempt-" + i);
+        }
+
```

Comment:
Suggest asserting that we have tripped the max-retries limit here:

```suggestion
        assertThat("replica should not be assigned", getReplica().state(), equalTo(ShardRoutingState.UNASSIGNED));
        assertThat("reroute should be a no-op", strategy.reroute(clusterState, "test"), sameInstance(clusterState));

```